upcloud/datasource_upcloud_storage.go

```diff
@@ -0,0 +1,172 @@
+			},
+			"name_regex": {
+				Description:      "Use regular expression to match storage name",
+				Type:             schema.TypeString,
+				Optional:         true,
+				ExactlyOneOf:     []string{"name", "name_regex"},
+				ValidateDiagFunc: validation.ToDiagFunc(validation.StringIsValidRegExp),
+			},
+			"name": {
+				Description:      "Exact name of the storage (same as title)",
+				Type:             schema.TypeString,
+				Optional:         true,
+				ExactlyOneOf:     []string{"name", "name_regex"},
+				ValidateDiagFunc: validation.ToDiagFunc(validation.StringLenBetween(0, 64)),
+			},
+			"zone": {
+				Description: "The zone in which the storage resides",
+				Type:        schema.TypeString,
+				Optional:    true,
+				Computed:    true,
+			},
+			"size": {
+				Description: "Size of the storage in gigabytes",
+				Type:        schema.TypeInt,
+				Computed:    true,
+			},
+			"state": {
+				Description: "Current state of the storage",
+				Type:        schema.TypeString,
+				Computed:    true,
+			},
+			"tier": {
+				Description: "Storage tier in use",
+				Type:        schema.TypeString,
+				Computed:    true,
+			},
+			"title": {
+				Description: "Title of the storage",
+				Type:        schema.TypeString,
+				Computed:    true,
+			},
+			"most_recent": {
+				Description: "If more than one result is returned, use the most recent storage",
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Default:     false,
+			},
+		},
+	}
+}
+
+func dataSourceUpCloudStorageRead(ctx context.Context, d *schema.ResourceData, meta interface{}) (diags diag.Diagnostics) {
+	svc := meta.(*service.Service)
+	var re *regexp.Regexp
+
+	nameRegex, nameRegexExists := d.GetOk("name_regex")
+	if nameRegexExists {
+		re = regexp.MustCompile(nameRegex.(string))
+	}
+
+	storages, err := svc.GetStorages(
+		&request.GetStoragesRequest{Type: d.Get("type").(string)})
+
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	name, nameExists := d.GetOk("name")
+	zone, zoneExists := d.GetOk("zone")
+	accessType, accessTypeExists := d.GetOk("access_type")
+	matches := make([]upcloud.Storage, 0)
+	for _, storage := range storages.Storages {
+		zoneMatch := (!zoneExists || zone == storage.Zone)
+		accessTypeMatch := (!accessTypeExists || accessType == storage.Access)
+		if nameExists && name == storage.Title && zoneMatch && accessTypeMatch {
+			matches = append(matches, storage)
+		} else if nameRegexExists && re != nil && re.MatchString(storage.Title) && zoneMatch && accessTypeMatch {
+			matches = append(matches, storage)
+		}
+	}
+	if len(matches) < 1 {
+		return diag.Errorf("query returned no results")
+	}
+
+	if len(matches) > 1 {
+		if !d.Get("most_recent").(bool) {
+			return diag.Errorf("query returned more than one result")
+		}
+		// sort storages by created timestamp
+		sort.Slice(matches, func(i, j int) bool {
+			return matches[i].Created.Unix() > matches[j].Created.Unix()
```

Comment:
This will not really work as expected I think. I don't know why `go-api-sdk` includes `Created` field for storage, but it does not seem to be populated (and it looks like it's intentional). So all storages will have the same created time, and that will be the default value to `time.Time{}`.

You can see this with this config for example:
```

data "upcloud_storage" "ubuntu_template" {
  type = "template"
  name_regex = "^Ubuntu"
  most_recent = true
}

resource "upcloud_server" "example" {
  hostname = "ubuntu-ecample"
  zone     = "pl-waw1"
  plan     = "1xCPU-1GB"

  template {
    storage = data.upcloud_storage.ubuntu_template.id
    size    = 25

    backup_rule {
      interval  = "daily"
      time      = "0100"
      retention = 8
    }
  }

  network_interface {
    type = "public"
  }

  network_interface {
    type = "utility"
  }
}
```

The server will be created with Ubuntu 18 template, even tho we have Ubuntu 20 available

Reply:
Thanks @ka-myl ! I'll see what is happening here. With custom images it seems to be working but I'll test it also to be sure.